src/app/layers/qgsapplayerhandling.cpp

```diff
@@ -1432,36 +1445,40 @@ T *QgsAppLayerHandling::addLayerPrivate( Qgis::LayerType type, const QString &ur
     }
     else
     {
-      result = qobject_cast< T * >( addSublayers( sublayers, name, QString(), addToLegend ).value( 0 ) );
+      const QList<QgsMapLayer *> layers { addSublayers( sublayers, name, QString(), addToLegend ) };
 
-      if ( result )
+      if ( ! layers.isEmpty() )
       {
         QString base( baseName );
         if ( settings.value( QStringLiteral( "qgis/formatLayerName" ), false ).toBool() )
         {
           base = QgsMapLayer::formatLayerName( base );
         }
-        result->setName( base );
+        for ( QgsMapLayer *layer : std::as_const( layers ) )
+        {
+          layer->setName( base );
+          result << qobject_cast<T *>( layer );
+        }
       }
     }
   }
   else
   {
     QgsMapLayerFactory::LayerOptions options( QgsProject::instance()->transformContext() );
     options.loadDefaultStyle = false;
-    result = qobject_cast< T * >( QgsMapLayerFactory::createLayer( uri, name, type, options, providerKey ) );
-    if ( result )
+    result.push_back( qobject_cast< T * >( QgsMapLayerFactory::createLayer( uri, name, type, options, providerKey ) ) );
+    if ( ! result.isEmpty() )
     {
       QString base( baseName );
       if ( settings.value( QStringLiteral( "qgis/formatLayerName" ), false ).toBool() )
       {
         base = QgsMapLayer::formatLayerName( base );
       }
-      result->setName( base );
-      QgsProject::instance()->addMapLayer( result, addToLegend );
+      result.first()->setName( base );
+      QgsProject::instance()->addMapLayer( result.first(), addToLegend );
 
-      QgisApp::instance()->askUserForDatumTransform( result->crs(), QgsProject::instance()->crs(), result );
-      QgsAppLayerHandling::postProcessAddedLayer( result );
+      QgisApp::instance()->askUserForDatumTransform( result.first()->crs(), QgsProject::instance()->crs(), result.first() );
+      QgsAppLayerHandling::postProcessAddedLayer( result.first() );
```

Comment:
shouldn't we iterate over all layers instead of just using the first one?

Reply:
I don't think so because in this branch canQuerySublayers is false and QgsMapLayerFactory::createLayer only returns a single layer, `results` will contain at most one single layer.

Reply:
ok, fair enough. perhaps add a comment comment to clarify this?

Reply:
done